cpp/ql/src/experimental/Security/CWE/CWE-193/ConstantSizeArrayOffByOne.ql

```diff
@@ -109,22 +110,39 @@ predicate isConstantSizeOverflowSource(Field f, PointerAddInstruction pai, int d
 
 module PointerArithmeticToDerefConfig implements DataFlow::ConfigSig {
   predicate isSource(DataFlow::Node source) {
-    isConstantSizeOverflowSource(_, source.asInstruction(), _)
+    isConstantSizeOverflowSource(_, _, source.asInstruction(), _)
   }
 
   pragma[inline]
   predicate isSink(DataFlow::Node sink) { isInvalidPointerDerefSink1(sink, _, _) }
 }
 
+module MergedPathGraph = DataFlow::MergePathGraph<PointerArithmeticToDerefFlow::PathNode, FieldAddressToPointerArithmeticFlow::PathNode, PointerArithmeticToDerefFlow::PathGraph, FieldAddressToPointerArithmeticFlow::PathGraph>;
```

Comment:
Using `DataFlow::MergePathGraph` is a great idea 🎉!

Reply:
It was just brought up somewhere else that another solution (instead of using two configurations) is to use a `FlowState`. Something like a single configuration that does:
```ql
module SingleConfiguration implements ... {
  predicate isSource(Node source, FlowState state) {
    state = 1 and PointerArithmeticToDerefFlow_isSource(source)
  }

  predicate isSink(Node sink, FlowState state) {
    state = 2 and FieldAddressToPointerArithmeticFlow_isSink(sink)
  }

  predicate isAdditionalFlowState(Node node1, FlowState state1, Node node2, FlowState state2) {
    state1 = 1 and
    PointerArithmeticToDerefFlow_isSink(state1)
    or
    state2 = 2 and
    FieldAddressToPointerArithmeticFlow_isSource(state1)
  }
}
```
That has several advantages:
1. We don't have to stitch to paths together
2. We keep the call context between the two phases of the query.

Reply:
The integer that represents array sizes can be bounded by "all integers used as sizes of  constant arrays", at least. For the `off-by-N` value we get similar binding issues as in the `cpp/overrun-write` query where we've currently bounded id somewhat arbitrarily as `[-32 ... 32]`.

Reply:
I was able to rephrase it using the initial array and the PointerAddInstruction as the state values, which sidesteps the binding problem entirely.